### PR TITLE
refactor(execute): update execute package to work with algorithm w

### PR DESCRIPTION
**Pull request by jsternberg**
### Done checklist
- [x] docs/SPEC.md updated
- [x] Test cases written

### CODE REVIEW

File: execute/row_fn.go [hunk 1]
```diff
@@ -13,52 +12,85 @@ import (
...
+		}
+		properties[i] = semantic.PropertyType{
+			Key:   []byte(c.Label),
+			Value: vtype,
+		}
 	}
+	return semantic.NewObjectType(properties), nil
 }
 
 func (f *dynamicFn) prepare(cols []flux.ColMeta, extraTypes map[string]semantic.MonoType) error {
```
Comment (jlapacik) [inline execute/row_fn.go]: Is extraTypes still necessary?

File: execute/row_fn.go [hunk 2]
```diff
@@ -13,52 +12,85 @@ import (
...
+	fn         *semantic.FunctionExpression
 	recordName string
-	record     *Record
 
-	recordCols map[string]int
-	references []string
+	// These are initialized after the function is prepared.
+	preparedFn compiler.Func
+	arg0       values.Object
+	args       values.Object
```
Comment (jlapacik) [inline execute/row_fn.go]: What are `args` and `args0`? Why do we need both?

File: execute/row_fn.go (hunk 1 above)
Comment (jsternberg) [reply execute/row_fn.go]: I think so. I think it's needed for reduce but I haven't gotten there yet honestly.

File: execute/row_fn.go (hunk 2 above)
Comment (jsternberg) [reply execute/row_fn.go]: `args` is the object for calling the function and `arg0` is the first argument specifically in the constructed object.

Arguments are sent to functions by using an object expression so `foo({a: 1, b: 2})` is two arguments, but the `Eval` function only takes the argument for the first one. So `arg0` is the object for `r` itself while `args` is for the all parameters.

**Pull request merged**